zebra-state/src/in_memory.rs

```diff
@@ -0,0 +1,119 @@
+use super::{Request, Response};
+use futures::prelude::*;
+use std::{
+    collections::{BTreeMap, HashMap},
+    error::Error,
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower::{buffer::Buffer, Service};
+use zebra_chain::{
+    block::{Block, BlockHeaderHash},
+    types::BlockHeight,
+};
+
+#[derive(Default)]
+struct ZebraState {
+    by_hash: HashMap<BlockHeaderHash, Arc<Block>>,
+    by_height: BTreeMap<BlockHeight, Arc<Block>>,
+}
+
+impl ZebraState {
+    fn insert(&mut self, block: impl Into<Arc<Block>>) {
+        let block = block.into();
+        let hash = block.as_ref().into();
+        let height = block.coinbase_height().unwrap();
+
+        assert!(
+            self.by_hash.insert(hash, block.clone()).is_none(),
+            "blocks shouldn't have the same hash"
```

Comment:
This assert will fail if insert is called twice, correct? Is there any reason we would want to disallow that?

Reply:
hmm, good point. this assert at least is okay, I was more worried about the second one which is just a guard against unimplemented features